### ISSUE TITLE
Upload file from web to remote USB disk that plug into PiKvm

### ISSUE THREAD

**Issue by KaoTuz** [pikvm/pikvm#1218]
**Is your feature request related to a problem? Please describe.**
Ans: No

                                                    
Remote Web<-Internet-> --> PiKvm<-->USB Disk>--> target PC 
USB disk  connect to PiKvm

We want to remote flash BIOS via our flash tool in UEFI shell
so we want to upload BIOS file in to remote usb disk. 
And control target PC boot into UEFI shell, then we execute flash command via usb disk

**Comment by Popkornium18** [comment]
You can create a disk image on your machine, create whatever filesystem you need in it (likely fat32), mount it, put the file into it, unmount it and then upload it like a normal disk image to the pikvm.

Look at step 2 of "Creating a simple fat32 image": https://docs.pikvm.org/msd/#writable-flash-drive

I successfully used that method to flash a BIOS over pikvm without actually plugging a USB stick into the system.

**Comment by KaoTuz** [comment]
Thanks for your reply

I have an usage scenarios
I am bios engineer
And i would need to copy bios binary to target PC (USB Disk) every time that i build an new bios
And boot the USB disk to UEFI shell (Not windows) than flash the BIOS

Does that means i need to un-mount and re-create new disk image when i need to add new file  to usb disk?

**Comment by Popkornium18** [comment]
With my proposed solution you would need to upload a new disk image every time, which might be too much of a hassle if you flash BIOSes multiple times every day. You could also try to mount the disk on the pikvm (writable) and scp the new bios file from your workstation onto the usb. Not sure if it works, but worth a try.

**Comment by mdevaev** [comment]
@KaoTuz Hello. Please help me understand your workflow and I'll figure out what can be done. It sounds super interesting. 

You have a test host that PiKVM is connected to, right? You are building a binary BIOS. Then you need to put it on some specialized USB flash drive, which has a utility for BIOS firmware. Then you connect it to the host, and the magic happens. Do I understand correctly?

**Comment by KaoTuz** [comment]
> @KaoTuz Hello. Please help me understand your workflow and I'll figure out what can be done. It sounds super interesting.
> 
> You have a test host that PiKVM is connected to, right? You are building a binary BIOS. Then you need to put it on some specialized USB flash drive, which has a utility for BIOS firmware. Then you connect it to the host, and the magic happens. Do I understand correctly?

Yes, i have a test host(remote target) that PiKVM is connect to.
## Standard Procedure to Flash BIOS
1. Copy bios file (Ex: Bios.bin ) to USB drive
2. Connect to test host
3. Reboot host
4. Boot to UEFI Shell (Not boot into windows or linux)
5. then it would auto execute batch file that i write to flash BIOS


## My Expected behavior of flashing BIOS via PIKvm
1. Copy bios file (Ex: Bios.bin ) to Pikvm Internal USB drive
- **Do not need to  re-make whole usb disk image, just add or override one file to the Internal USB drive that existed**
3. Re-connect internal USB drive to host
4. Reboot host
5. Boot to UEFI Shell (Not boot into windows or linux)
6. then it would auto execute batch file that i write to flash BIOS


Due to i would build a lot of bios to testing host, so i may need to copy bios file to usb driver frequently
So , i just want to add or override one file to the Internal  USB drive to reduce time

**Comment by mdevaev** [comment]
I've more or less figured out what you need. I need to think about the implementation.

**Comment by steverweber** [comment]
hit a similar issues here was my hacky workaround.

Use 256mb RAM disk
--------------------------

```
[root@pikvm ~]# 

# hack in change to allow drives from symlinks
# the symlink points to an image in /tmp
rw
    vi /usr/lib/python3.12/site-packages/kvmd/plugins/msd/otg/storage.py
    #    elif item.is_file(follow_symlinks=True):
ro
# reboot for the storage.py change
reboot

...

# helper script to build a ramdisk
<<EOF cat > /usr/bin/256mb-ramdisk
#!/usr/bin/env bash
dd if=/dev/zero of=/tmp/256mb-ramdisk.img bs=1M count=256 status=progress
mkfs.vfat -I -F 32 -n 'RAMDISK' /tmp/256mb-ramdisk.img
chmod 666 /tmp/256mb-ramdisk.img
test -e /var/lib/kvmd/msd/256mb-ramdisk.img || (
    kvmd-helper-otgmsd-remount rw
    ln -sT /tmp/256mb-ramdisk.img /var/lib/kvmd/msd/256mb-ramdisk.img
    kvmd-helper-otgmsd-remount ro
)
EOF
chmod +x /usr/bin/256mb-ramdisk

# hacky script to remount the ramdisk after guest makes changes to files
<<EOF cat > /usr/bin/256mb-ramdisk-remount
#!/usr/bin/env bash
mkdir -p /tmp/256mb-ramdisk
umount /tmp/256mb-ramdisk || true
mount /tmp/256mb-ramdisk.img /tmp/256mb-ramdisk || true
echo /tmp/256mb-ramdisk
ls /tmp/256mb-ramdisk
EOF
chmod +x /usr/bin/256mb-ramdisk-remount

```

How to use:

1. as root do the above 
2. from pikvm terminal as root run: 256mb-ramdisk
3. add drive to guest as flash RW.. add files...
4. from pikvm terminal root: 256mb-ramdisk-remount
5. grab files from /tmp/256mb-ramdisk

to send files... add files to /tmp/256mb-ramdisk then attach the storage to the guest.

Notes:
- after guest changes files need to remount using 256mb-ramdisk-remount to see them
- on a windows XP guest after connecting drive as RW flash I need to reboot the guest to see the drive as a RW disk not a CDROM.
- piKVM says "current image out of storage"... perhaps safe to ignore

**Comment by steverweber** [comment]
hacky but worked for me.

256mb RAM disk
--------------

```
# script to build ramdisk
<<EOF cat > /usr/bin/256mb-ramdisk-recreate
#!/usr/bin/env bash
dd if=/dev/zero of=/tmp/256mb-ramdisk.img bs=1M count=256 status=progress
mkfs.vfat -I -F 32 -n 'RAMDISK' /tmp/256mb-ramdisk.img
EOF
chmod +x /usr/bin/256mb-ramdisk-recreate

# hacky script to remount the ramdisk.. run this after guest or pikvm makes file changes
<<EOF cat > /usr/bin/256mb-ramdisk-remount
#!/usr/bin/env bash
kvmd-otgmsd -i 0 --eject
sleep 1
kvmd-otgmsd -i 0 --set-rw 1 --set-cdrom 0 --set-image /tmp/256mb-ramdisk.img
mkdir -p /tmp/256mb-ramdisk
umount /tmp/256mb-ramdisk || true
mount /tmp/256mb-ramdisk.img /tmp/256mb-ramdisk || true
echo /tmp/256mb-ramdisk
EOF
chmod +x /usr/bin/256mb-ramdisk-remount
```

How to use

1. from pikvm terminal as root run: 256mb-ramdisk-recreate and 256mb-ramdisk-remount
2. guest adds files... or pikvm adds files to /tmp/256mb-ramdisk
3. from pikvm terminal as root run: 256mb-ramdisk-remount
4. the guest or pikvm grabs the new files

Notes:
- on a windows XP guest after connecting drive as RW flash I needed to reboot the guest to see the drive as a RW disk.
- piKVM says "current image out of storage"... perhaps safe to ignore
- Perhaps there is better mount option that wont require the remounting; if yes please share

**Comment by dragon788** [comment]
> Due to i would build a lot of bios to testing host, so i may need to copy bios file to usb driver frequently So , i just want to add or override one file to the Internal USB drive to reduce time that re-build whole disk

Depending on how much storage you have and historical versions you might need to keep you could use the API to upload a new image that contains your "environment" and the update (FreeDOS + files ftw?).

You could potentially use the "multiple" MSD option to have a primary image that is just FreeDOS that doesn't change, and then swap out the secondary using the CLI or API to be an "image" that is a FAT32 filesystem containing your update files (and I would hope your CI/CD build system could manage making those images pretty easily).

https://docs.pikvm.org/api/#upload-msd-image
